File: Feliz.PigeonMaps/Marker.js
import { Component, createElement } from 'react'

class Marker extends Component {

    constructor(props) {
        super(props)

        this.state = {
            hover: false
        }
    }

    // what do you expect to get back with the event
    eventParameters(event) {
        return {
            anchor: this.props.anchor
        }
    }

    // controls
    isRetina() {
        return typeof window !== 'undefined' && window.devicePixelRatio >= 2
    }

    // modifiers
    isHover() {
        return typeof this.props.hover === 'boolean' ? this.props.hover : this.state.hover
    }

    // delegators

    handleClick(event) {
        this.props.onClick && this.props.onClick(this.eventParameters(event))
    }

    handleContextMenu(event) {
        this.props.onContextMenu && this.props.onContextMenu(this.eventParameters(event))
    }

    handleMouseOver(event) {
        this.props.onMouseOver && this.props.onMouseOver(this.eventParameters(event))
        this.setState({ hover: true })
    }

    handleMouseOut(event) {
        this.props.onMouseOut && this.props.onMouseOut(this.eventParameters(event))
        this.setState({ hover: false })
    }

    // render

    render() {
        const { left, top } = this.props

        const style = {
            position: 'absolute',
            transform: `translate(${left - (this.props.offsetLeft || 0)}px, ${top - (this.props.offsetTop || 0)}px)`
        }

        return (
            createElement('div', {
                style: style,
                onClick: this.handleClick.bind(this),
                onContextMenu: this.handleContextMenu.bind(this),
                onMouseOver: this.handleMouseOver.bind(this),
                onMouseOut: this.handleMouseOut.bind(this),
                children: this.props.render && this.props.render({ hovered: this.state.hover, anchor: this.props.anchor })
            })
        )
    }
}


export const createMarker = function (extraProps) {
    return createElement(Marker, { ...extraProps })
}

File: Feliz.PigeonMaps/Marker.fs
namespace Feliz.PigeonMaps

open Feliz
open Fable.Core
open Fable.Core.JsInterop

[<Erase>]
type marker =
    static member inline anchor(latitude: float, longitude: float) =
        Interop.mkAttr "anchor" [| latitude; longitude |]
    static member inline offsetLeft (offset: int) =
        Interop.mkAttr "offsetLeft" offset
    static member inline offsetTop (offset: int) =
        Interop.mkAttr "offsetTop" offset
    static member inline onClick (handler: MarkerClickEventArgs -> unit) =
        Interop.mkAttr "onClick" handler
    static member inline onMouseOver (handler: MarkerClickEventArgs -> unit) =
        Interop.mkAttr "onMouseOver" handler
    static member inline onMouseOut (handler: MarkerClickEventArgs -> unit) =
        Interop.mkAttr "onMouseOut" handler
    static member inline onContextMenu (handler: MarkerClickEventArgs -> unit) =
        Interop.mkAttr "onContextMenu" handler
    static member inline render (handler: IMarkerRenderProperties -> Fable.React.ReactElement) =
        Interop.mkAttr "render" handler
    static member inline render (handler: IMarkerRenderProperties -> Fable.React.ReactElement list) =
        Interop.mkAttr "render" (handler >> React.fragment)

[<Erase>]
type marker<'t> =
    static member inline anchor(latitude: float, longitude: float) =
        Interop.mkAttr "anchor" [| latitude; longitude |]
    static member inline payload (value: 't) =
        Interop.mkAttr "payload" value
    static member inline onClick (handler: MarkerClickEventArgs<'t> -> unit) =
        Interop.mkAttr "onClick" handler
    static member inline onMouseOver (handler: MarkerClickEventArgs<'t> -> unit) =
        Interop.mkAttr "onMouseOver" handler
    static member inline onMouseOut (handler: MarkerClickEventArgs<'t> -> unit) =
        Interop.mkAttr "onMouseOut" handler
    static member inline onContextMenu (handler: MarkerClickEventArgs<'t> -> unit) =
        Interop.mkAttr "onContextMenu" handler
    static member inline render (handler: IMarkerRenderProperties -> Fable.React.ReactElement) =
        Interop.mkAttr "render" handler
    static member inline render (handler: IMarkerRenderProperties -> Fable.React.ReactElement list) =
        Interop.mkAttr "render" (handler >> React.fragment)
    static member inline offsetLeft (offset: int) =
        Interop.mkAttr "offsetLeft" offset
    static member inline offsetTop (offset: int) =
        Interop.mkAttr "offsetTop" offset